core/src/main/java/org/infinispan/container/versioning/irac/DefaultIracTombstoneManager.java

```diff
@@ -232,117 +287,173 @@ public boolean contains(IracTombstoneInfo tombstone) {
+         return iracManager.running().containsKey(tombstone.getKey()) ? Action.KEEP_TOMBSTONE : Action.NOTIFY_PRIMARY_OWNER;
+      } else {
+         // primary owner, check all remote sites.
+         return iracManager.running().containsKey(tombstone.getKey()) ? Action.KEEP_TOMBSTONE : Action.CHECK_REMOTE_SITE;
       }
+   }
 
-      CompletionStage<Void> checkRemoteSites() {
+   private final class CleanupTask implements Function<Void, CompletionStage<Void>>, Runnable {
+      private final Collection<IracTombstoneInfo> tombstoneToCheck;
+      private final IntSet tombstoneToKeep;
+      private final int id;
+      private volatile boolean failedToCheck;
+
+      private CleanupTask(Collection<IracTombstoneInfo> tombstoneToCheck) {
+         this.tombstoneToCheck = tombstoneToCheck;
+         tombstoneToKeep = IntSets.concurrentSet(tombstoneToCheck.size());
+         failedToCheck = false;
+         id = tombstoneToCheck.hashCode();
+      }
+
+      Flowable<Void> check() {
+         if (log.isTraceEnabled()) {
+            log.tracef("[cleanup-task-%d] Running cleanup task with %s tombstones to check", id, tombstoneToCheck.size());
+         }
+         if (tombstoneToCheck.isEmpty()) {
+            return Flowable.empty();
+         }
+         List<Object> keys = tombstoneToCheck.stream()
+               .map(IracTombstoneInfo::getKey)
+               .collect(Collectors.toList());
+         IracTombstoneRemoteSiteCheckCommand cmd = commandsFactory.buildIracTombstoneRemoteSiteCheckCommand(keys);
          // if one of the site return true (i.e. the key is in updateKeys map, then do not remove it)
-         AggregateCompletionStage<Boolean> stage = CompletionStages.orBooleanAggregateCompletionStage();
+         AggregateCompletionStage<Void> stage = CompletionStages.aggregateCompletionStage();
          for (XSiteBackup backup : asyncBackups) {
             if (takeOfflineManager.getSiteState(backup.getSiteName()) == SiteState.OFFLINE) {
                continue; // backup is offline
             }
             // we don't need the tombstone to query the remote site
-            IracTombstoneRemoteSiteCheckCommand cmd = commandsFactory.buildIracTombstoneRemoteSiteCheckCommand(tombstone.getKey());
-            stage.dependsOn(rpcManager.invokeXSite(backup, cmd));
+            stage.dependsOn(rpcManager.invokeXSite(backup, cmd).thenAccept(this::mergeIntSet));
          }
          // in case of exception, keep the tombstone
-         return stage.freeze()
-               .exceptionally(toTrueFunction())
-               .thenComposeAsync(this, blockingExecutor);
+         return Flowable.fromCompletable(
+               Completable.fromCompletionStage(
+                     stage.freeze()
+                           .exceptionally(this::onException)
+                           .thenComposeAsync(this, blockingManager.asExecutor("tombstone-xsite-response"))));
```

Comment:
We still need to discuss what needs to be ran on the blocking thread. Unfortunately, using the blocking executor can still leak blocking threads, so it should be used infrequently.

I can suggest what to do here, but need to know what part actually needs to run on the blocking thread.